semgrep-core/src/analyzing/Naming_AST.ml

```diff
@@ -430,16 +433,23 @@ let resolve2 lang prog =
           k x
         )
       );
-      V.kdef = (fun (k, _v) x ->
+      V.kdef = (fun (k, v) x ->
         match x with
-        | { name = EN (Id (id, id_info)); _},
+        | { name = EN (Id (id, id_info)); _} as ent,
           (* note that some languages such as Python do not have VarDef
            * construct
            * todo? should add those somewhere instead of in_lvalue detection? *)
           VarDef ({ vinit; vtype }) when is_resolvable_name_ctx env lang ->
-            (* Need to visit expressions first so that type is populated *)
-            (* If we do var a = 3, then var b = a, we want to propagate the type of a *)
-            k x;
+            (* Need to visit expressions first so that type is populated, e.g.
+             * if we do var a = 3, then var b = a, we want to propagate the type of a. *)
+            (* Note that we are careful to visit each part of the declaration
+             * separately in order to avoid the v_vardef_as_assign_expr
```

Comment:
yep, we need to remove the v_vardef_as_assign_expr in the visitor; it's too subtle. We need to move the logic in generic_vs_generic.ml and perform the equivalence there.

Reply:
Let me look into that next week.